File: python/otel_demo_devoxx.py
import random
import time
from opentelemetry import trace, metrics
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor, OTLPSpanExporter
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader, OTLPMetricExporter

# Traces
trace.set_tracer_provider(TracerProvider())
tracer = trace.get_tracer(__name__)
span_exporter = OTLPSpanExporter(endpoint="http://localhost:4317", insecure=True)
trace.get_tracer_provider().add_span_processor(BatchSpanProcessor(span_exporter))

# Metrics
meter_provider = MeterProvider(
    metric_readers=[PeriodicExportingMetricReader(OTLPMetricExporter(endpoint="http://localhost:4317", insecure=True))]
)
metrics.set_meter_provider(meter_provider)
meter = metrics.get_meter(__name__)
counter = meter.create_counter("demo_counter")

print("Demo running. Sending traces and metrics to OTEL Viewer at localhost:4317...")

for i in range(10):
    with tracer.start_as_current_span(f"operation-{i}"):
        counter.add(i, {"iteration": i})
        time.sleep(0.5)

print("Demo finished.")

File: go/otel_demo_devoxx.go
package main

import (
    "context"
    "fmt"
    "math/rand"
    "time"

    "go.opentelemetry.io/otel"
    "go.opentelemetry.io/otel/attribute"
    "go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
    "go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
    "go.opentelemetry.io/otel/sdk/metric"
    sdktrace "go.opentelemetry.io/otel/sdk/trace"
)

func main() {
    ctx := context.Background()

    // Tracer
    traceExporter, _ := otlptracegrpc.New(ctx)
    tp := sdktrace.NewTracerProvider(
        sdktrace.WithBatcher(traceExporter),
    )
    otel.SetTracerProvider(tp)
    tracer := otel.Tracer("demo-tracer")

    // Metrics
    metricExporter, _ := otlpmetricgrpc.New(ctx)
    mp := metric.NewMeterProvider(
        metric.WithReader(metric.NewPeriodicReader(metricExporter, metric.WithInterval(time.Second))),
    )
    meter := mp.Meter("demo-meter")
    counter, _ := meter.Int64Counter("demo_counter")

    fmt.Println("Demo running. Sending traces and metrics to OTEL Viewer at localhost:4317...")

    for i := 0; i < 10; i++ {
        _, span := tracer.Start(ctx, fmt.Sprintf("operation-%d", i))
        counter.Add(ctx, int64(i), attribute.Int("iteration", i))
        span.End()
        time.Sleep(500 * time.Millisecond)
    }

    fmt.Println("Demo finished.")
}
